src/functions.php

```diff
@@ -478,23 +478,48 @@ function is_settled(PromiseInterface $promise)
 function coroutine(callable $generatorFn)
 {
     $generator = $generatorFn();
-    return __next_coroutine($generator->current(), $generator)->then();
+    $state = new \stdClass;
```

Comment:
I think the proposed change would be easier to read if you called three named methods (e.g., `clearCurrentPromise`, `setCurrentPromise`, and `rejectCoroutinePromise`) on a small class that encapsulated the two promise properties. Additionally, it would be easier to test in isolation, which is important in a project that accepts pull requests from the community. I would need to see a performance test to gauge which approach is faster.